File: 0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.cpp
class Solution {
public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        // For priority_queue, comparator defines when one element should be 
        // considered lower priority than another. 
        
        // a has lower priority than b if dist(a) > dist(b)
        auto comp = [](vector<int> a, vector<int> b) {
            long long distA = a[0] * a[0] + a[1] * a[1];
            long long distB = b[0] * b[0] + b[1] * b[1];

            return distA > distB;
        };

        
        priority_queue<vector<int>, vector<vector<int>>, decltype(comp)> minHeap;
        vector<vector<int>> res;

        for (auto& pts : points) {
            minHeap.push(pts);
        }


        while (k--) {
            res.push_back(minHeap.top());
            minHeap.pop();

        }

        return res;


        
    }
};

File: 0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.py
class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:

        pq = []
        output = []

        for x, y in points:
            dist = x ** 2 + y ** 2
            heapq.heappush(pq, [dist, x, y])
        
        while k > 0:
            dist, x, y = heapq.heappop(pq)
            output.append([x, y])
            k -= 1
        
        return output


        